### PR TITLE
Use OP blob encoding scheme with zlib compression

### PR DESCRIPTION
**Pull request by smtmfft**
How OP encode blob: https://www.notion.so/taikoxyz/How-OP-encodes-blob-8096cad7765d48cdab9667722330aef0

### CODE REVIEW

File: lib/src/taiko/host.rs [hunk 1]
```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
...
+        Err(_) => return Vec::new(),
+    };
+
+    // check the version
+    if blob_buf[BLOB_VERSION_OFFSET] != BLOB_ENCODING_VERSION {
+        return Vec::new();
+    }
+
+    // decode the 3-byte big-endian length value into a 4-byte integer
+    let output_len: usize =
```
Comment (johntaiko) [inline lib/src/taiko/host.rs:218]:
```suggestion
    let output_len =
```

File: lib/src/taiko/host.rs [hunk 2]
```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
...
+        ((blob_buf[2] as u32) << 16 | (blob_buf[3] as u32) << 8 | (blob_buf[4] as u32)) as usize;
+    if output_len > MAX_BLOB_DATA_SIZE {
+        return Vec::new();
+    }
+
+    // round 0 is special cased to copy only the remaining 27 bytes of the first field element
+    // into the output due to version/length encoding already occupying its first 5 bytes.
+    let mut output: [u8; MAX_BLOB_DATA_SIZE] = [0; MAX_BLOB_DATA_SIZE];
```
Comment (johntaiko) [inline lib/src/taiko/host.rs:226]:
```suggestion
    let mut output = [0; MAX_BLOB_DATA_SIZE];
```

File: lib/src/taiko/host.rs [hunk 3]
```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
...
+    output[0..27].copy_from_slice(&blob_buf[5..32]);
+
+    // now process remaining 3 field elements to complete round 0
+    let mut opos: usize = 28; // current position into output buffer
+    let mut ipos: usize = 32; // current position into the input blob
+    let mut encoded_byte: [u8; 4] = [0; 4]; // buffer for the 4 6-bit chunks
+    encoded_byte[0] = blob_buf[0];
+    for i in 1..4 {
+        (encoded_byte[i], opos, ipos) =
+            match decode_field_element(&blob_buf, opos, ipos, &mut output) {
```
Comment (johntaiko) [inline lib/src/taiko/host.rs:236]:
```suggestion
            decode_field_element(&blob_buf, opos, ipos, &mut output).unwrap_or_default();
```

File: lib/src/taiko/host.rs [hunk 4]
```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
...
-        decoded_len += segment_len;
-        chunk.push(segment.to_vec());
+const BLOB_DATA_CAPACITY: usize = BLOB_FIELD_ELEMENT_NUM * BLOB_FIELD_ELEMENT_BYTES;
+const BLOB_VERSION_OFFSET: usize = 1;
+const BLOB_ENCODING_VERSION: u8 = 0;
+const MAX_BLOB_DATA_SIZE: usize = (4 * 31 + 3) * 1024 - 4;
+
+// decoding https://github.com/ethereum-optimism/optimism/blob/develop/op-service/eth/blob.go
+fn decode_blob_data(blob_str: &str) -> Vec<u8> {
+    let blob_buf: Vec<u8> = match hex::decode(blob_str.to_lowercase().trim_start_matches("0x")) {
```
Comment (johntaiko) [inline lib/src/taiko/host.rs:207]:
```suggestion
    let blob_buf = match hex::decode(blob_str.to_lowercase().trim_start_matches("0x")) {
```

File: lib/src/taiko/host.rs (hunk 1 above)
Comment (smtmfft) [reply lib/src/taiko/host.rs]: fixed

File: lib/src/taiko/host.rs (hunk 2 above)
Comment (smtmfft) [reply lib/src/taiko/host.rs]: Fixed

**Review by johntaiko — APPROVED** [review]

File: lib/src/taiko/host.rs [hunk 5]
```diff
@@ -197,37 +197,115 @@ fn execute_data<N: NetworkStrategyBundle<TxEssence = EthereumTxEssence>>(
...
+    if blob_buf[BLOB_VERSION_OFFSET] != BLOB_ENCODING_VERSION {
+        return Vec::new();
+    }
+
+    // decode the 3-byte big-endian length value into a 4-byte integer
+    let output_len =
+        ((blob_buf[2] as u32) << 16 | (blob_buf[3] as u32) << 8 | (blob_buf[4] as u32)) as usize;
+    if output_len > MAX_BLOB_DATA_SIZE {
+        return Vec::new();
+    }
```
Comment (Brechtpd) [inline lib/src/taiko/host.rs:222]: So version is at byte 1 and length is only 3 bytes? If so, slightly different from the figure in the notion I think. Though avoiding byte 0 makes sense, even though you could argue the version could be stored in just 6 bits, but doesn't matter, just trying to understand and used your figure to do that!

Comment (smtmfft) [reply lib/src/taiko/host.rs:222]: oh yes, by design the version & len byte is buf[1..5], not [0..5], I will update the pic.

**Review by Brechtpd — APPROVED** [review]

**Comment by smtmfft** [comment]
Added zlib support, it seems quite light weight comparing to kzg commit, 2 seg vs 200 seg (2^20 cycle each seg).

**Comment by Brechtpd** [comment]
One thing we're going to need to verify is the worst case scenario, similar to RLP data. Because if someone is able to generate data that is somehow very large decoded (either correctly or incorrectly) it may very easily be impossible to prove this data stream because of massive memory or processing requirements. Or it may even still be possible to prove it in practice, but the cost could be extremely large so it could be used to attack the protocol (cost to generate proof > proof bond).

**Comment by smtmfft** [comment]
> attack the protocol (cost to generate proof > proof bond).

I think we can add some cost functions in host to pre-evaluate the VM execution, but a conflicting thing here is that tier selection is random on chain, it may force prover to build ZK proof. Or we make those limits part of consensus? (i.e., impl same cost functions in client)

**Comment by Brechtpd** [comment]
> > attack the protocol (cost to generate proof > proof bond).
> 
> I think we can add some cost functions in host to pre-evaluate the VM execution, but a conflicting thing here is that tier selection is random on chain, it may force prover to build ZK proof. Or we make those limits part of consensus? (i.e., impl same cost functions in client)

yeah those limitations will have to be implemented in the client/consensus part as well, they need to be consistent and produce the same result in all cases because we'll have to have zk proofs for all blocks pretty soon if not at mainnet.

And I guess this could even be an issue in the SGX prover.

**Pull request merged**